core/controllers/library.py

```diff
@@ -45,16 +46,14 @@ def get_matching_activity_dicts(query_string, search_cursor):
     if not search_cursor:
         collection_ids, _ = (
             collection_services.get_collection_ids_matching_query(
-                query_string))
+                query_string, categories, language_codes))
```

Comment:
Through this PR we are manipulating the query_strings in 3+ places, if we are going to revisit and work on elstic_search_service in the future then why not just change elastic_search_services.search function such that it extracts required info from the query_string and makes the query? is there any issue with just changing the search function?

Reply:
What do you mean by "Through this PR we are manipulating the query_strings in 3+ places"? My intention was to have this all sorted out in the library controller and then we'll have 3 separate args passed through the rest of the codebase with clear conceptual identity.

The previous query string format was built based on the formatting needs of the GAE search services implementation, so I don't see a strong need to keep it.

Reply:
I mean we are doing the following in this PR:
1. Extracting categories and language_code in the controller.
2. Changing the gae search query method to regenerate the query_string using extracted categories and langauge_codes
3. Changing the elasticsearch_service.search method to use the categories/lanague_code to generate the query.

Instead, I'm suggesting whether it will be possible to just change the elastic_serach_service.search method to use the old query_string and extract the categories and language_codes from the query_string and generate the new query JSON as needed? (This way we will only be changing one function.)

Reply:
Oh, I see, thanks for clarifying. No, I don't think that's the right thing to do conceptually, because it locks us into using the old query_string format which was optimized for the gae implementation which is going away. Does that make things clearer?

Reply:
As per [this comment](https://github.com/oppia/oppia/pull/11500#discussion_r549085918) we are going to revisit the search service and currently, so I thought we should do minimal changes as of now. 

> because it locks us into using the old query_string format which was optimized for the gae implementation which is going away.

In that case, should we just change the frontend to send categories and language_codes as a list instead of string with "OR" (which is optimized for gae)?

Reply:
Ah -- that principle works as long as the minimal changes don't need to be reverted in the future. In this case, they will, so I don't want to code something that's actually wrong in the first place unless there's a compelling reason otherwise.

I think I am fine with leaving the current frontend code as-is. It is a bit tricky to pass a list through GET.